x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/NativeController.java

```diff
@@ -152,4 +200,75 @@ public void stop() throws IOException {
         // The C++ process will exit when it gets EOF on the command stream
         commandStream.close();
     }
+
+    private void setupResponseTracker(int commandId) {
+        ResponseTracker tracker = new ResponseTracker();
+        ResponseTracker previous = responseTrackers.put(commandId, tracker);
+        assert previous == null;
+    }
+
+    private void removeResponseTracker(int commandId) {
+        responseTrackers.remove(commandId);
+    }
+
+    private void awaitCompletion(int commandId) throws IOException, InterruptedException {
+
+        ResponseTracker ourResponseTracker = responseTrackers.get(commandId);
+        assert ourResponseTracker != null;
+
+        // If our response has not been seen already (by another thread), parse messages under lock until it is seen.
+        // This approach means that of all the threads waiting for controller responses, one is parsing the messages
+        // on behalf of all of them, and the others are blocked.  When the thread that is parsing gets the response
+        // it needs another thread will pick up the parsing.
+        if (ourResponseTracker.hasResponded() == false) {
+            synchronized (responseIteratorHolder) {
```

Comment:
I am curious on locking on this and not on `ourResponseTracker` for the entire while loop.

If we lock on this, that means that any other command could not be processed. 

Our commands are always handled in a single thread and could NEVER be processed out of order correct? If they are ever processed out of order, this seems like it would cause a thread lock.

Reply:
It's correct that at the moment the C++ always processes commands in order.  This may not always be the case though.

It's also true that in the future if two commands were processed out-of-order in the C++ then this Java code would delay one of the threads waiting for a response unnecessarily.  However, it wouldn't be a deadlock - it would just mean threads 1 and 2 would unblock on the response to command 1 (or disconnection of `controller`) even though thread 2 could have unblocked on the earlier response to command 2.

So some optimisation work could be done here if the C++ is ever modified to process commands out-of-order.

The actual parsing needs to synchronize on `responseIteratorHolder`, because only one thread can be parsing the stream at any time.  What _could_ be done is that the other threads simply wait for their response latches on the assumption that the thread doing the parsing will count these down.  _However_, this is non-trivial, because if the thread doing the parsing has just decided that it can stop parsing but hasn't exited the synchronized block then _no_ thread would be parsing.  So it would need more complex synchronization to ensure that the decision of one thread to wait for its latch instead of for the parser lock would need to be atomic with the thread currently holding the parser lock to decide to stop parsing (which is _not_ the same as releasing the parser lock).  Rather than risk messing this up I went for the simpler approach.

The alternative is to have a dedicated thread for parsing responses like we do for most C++ outputs, but this seems wasteful in the case of `controller` where there is very low traffic and we'd be permanently hogging a thread for it.

Reply:
> because only one thread can be parsing the stream at any time.

👍 

Thanks for the additional information. It just wasn't obvious to me :D